rust/projects/file-io/src/kv.rs

```diff
@@ -63,31 +59,137 @@ impl KvStore {
+        }
         Ok(())
     }
 
     /// Gets the string value of a given string key.
     ///
     /// Returns `None` if the given key does not exist.
-    pub fn get(&self, key: String) -> Result<Option<String>> {
-        Ok(self.map.get(&key).cloned())
+    pub fn get(&mut self, key: String) -> Result<Option<String>> {
+        self.kv_log.get(key)
+    }
+
+    fn compact(&mut self) -> Result<()> {
+        // The new log file for merged entries
+        let tmp_log_path = self.path.join("data.log.new");
+        let mut new_writer = BufWriter::new(
+            OpenOptions::new()
+                .create(true)
+                .truncate(true)
+                .write(true)
+                .open(&tmp_log_path)?,
+        );
+
+        // Copy content in order to reduce seeks
+        let mut pos_vec: Vec<_> = self.kv_log.index.iter().collect();
+        pos_vec.sort_unstable_by_key(|(_, cmd_pos)| cmd_pos.pos);
+        let mut new_pos = 0; // pos in the new log file
+        let mut new_index = HashMap::new(); // index map for the new log file
+        for (key, cmd_pos) in pos_vec {
+            if self.kv_log.reader.pos != cmd_pos.pos {
+                self.kv_log.reader.seek(SeekFrom::Start(cmd_pos.pos))?;
+            }
+
+            let mut entry_reader = (&mut self.kv_log.reader).take(cmd_pos.len);
+            let len = io::copy(&mut entry_reader, &mut new_writer)?;
+            new_index.insert(key.clone(), (new_pos..new_pos + len).into());
+            new_pos += len;
+        }
+        drop(new_writer);
+
+        // As all entries are written to the log, we can safely rename it to a valid log file name
+        let log_path = new_log_path(&self.path);
+        fs::rename(tmp_log_path, &log_path)?;
+
+        // Reopen using the new file name
+        let mut kv_log = KvLog::open(&log_path)?;
+        // Use the index map built on writing instead of reloading the log file
+        kv_log.index = new_index;
+        // Update the KvLog we are using
+        mem::swap(&mut self.kv_log, &mut kv_log);
+
+        // Close old log file before removing it. (It's a must on Windows I think)
+        let old_path = kv_log.path.clone();
+        drop(kv_log);
+        fs::remove_file(old_path)?;
+
+        Ok(())
+    }
+}
+
+struct KvLog {
+    path: PathBuf,
+    reader: BufReaderWithPos<File>,
+    writer: BufWriterWithPos<File>,
+    // stores keys and the pos of the last command to modify each
+    index: HashMap<String, CommandPos>,
+    uncompacted: u64,
+}
+
+impl KvLog {
+    // Pay attention that it does not load the log file automatically
+    fn open(path: impl Into<PathBuf>) -> Result<Self> {
+        let path = path.into();
+        let mut writer = BufWriterWithPos::new(
+            OpenOptions::new()
+                .create(true)
+                .read(true)
+                .append(true)
+                .open(&path)?,
+        )?;
+        // Set pos to end of file
+        writer.seek(SeekFrom::End(0))?;
```

Comment:
If this file handle is only doing writes then I don't think the seek is necessary.

Reply:
This seek is necessary. 

The file cursor is stored as `pos` in `BufferedWriterWithPos`, and we need to save that pos to the index for the following get commands.

Because the file mode is set to append, the pos we save is not synced with the actual position the write occurs. This seek just sets the pos in  `BufferedWriterWithPos` to make `pos` correct.

I'll update the comment here and modify the test to fail the case if this seek is missing. (I think other implementors are unlikely to wrap a writer with pos, though)

Reply:
You are correct. Thanks.